File: csharp/lesson 10/UniquePathsRecursive.cs
// https://leetcode.com/problems/unique-paths/

public class Solution
{
    public int UniquePaths(int m, int n)
    {
        var dp = new int[m + 1][];
        for (var i = 0; i < dp.Length; i++) dp[i] = new int[n + 1];
        
        dp[1][1] = 1;
        return UniquePaths(m, n, dp);
    }
    
    private int UniquePaths(int row, int col, int[][] dp)
    {
        if (row < 0 || col < 0) return 0;
        if (dp[row][col] > 0) return dp[row][col];
        
        dp[row][col] = UniquePaths(row - 1, col, dp) + UniquePaths(row, col - 1, dp);
        return dp[row][col];
    }
}

File: kotlin/lesson 10/UniquePathsRecursive.kt
// https://leetcode.com/problems/unique-paths/

class Solution {
    fun uniquePaths(m: Int, n: Int): Int {
        var dp = Array(m + 1, { i -> Array(n + 1, { j -> 0 }) })
        dp[1][1] = 1
        return uniquePaths(m, n, dp);
    }
    
    
    fun uniquePaths(row: Int, col: Int, dp: Array<Array<Int>>): Int {
        if (row < 0 || col < 0) return 0
        if (dp[row][col] > 0) return dp[row][col]
        
        dp[row][col] = uniquePaths(row - 1, col, dp) + uniquePaths(row, col - 1, dp)
        return dp[row][col]
    }
}
